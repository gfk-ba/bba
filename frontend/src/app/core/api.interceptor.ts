import { Injectable } from '@angular/core';
import { Reset } from '@dcs/ngx-tools';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { catchError, flatMap, map, take } from 'rxjs/operators';
import { State } from '../reducers';
import { tokenSelector } from '../reducers/login/login.selectors';
import {
  HttpEvent,
  HttpHandler,
  HttpInterceptor,
  HttpRequest,
  HttpErrorResponse,
} from '@angular/common/http';

@Injectable()
export class ApiInterceptor implements HttpInterceptor {
  constructor(private store: Store<State>) {}

  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    return this.store.pipe(
      take(1),
      map(state => tokenSelector(state)),
      flatMap(token => {
        request = request.clone({
          setHeaders: {
            Authorization: `Bearer ${token.accessToken}`,
          },
        });
        return next.handle(request).pipe(
          catchError((e: HttpErrorResponse) => {
            if (e.status === 401) {
              this.store.dispatch(new Reset());
            }
            throw e;
          })
        );
      })
    );
  }
}
