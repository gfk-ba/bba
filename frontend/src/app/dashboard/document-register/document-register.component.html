<div class="box">
  <form [formGroup]="form" (submit)="submitForm($event)">
    <div class="row">

      <div class="col-md-7 left-part">
        <div class="row">
          <div class="col-md-6 part">
            <div class="flex-wrapper">
              <select formControlName="projectId" class="form-control">
                <option [value]="null">Select project</option>
                <option *ngFor="let project of projects" [value]="project.id">
                  {{ project.name }}
                </option>
              </select>
              <button type="button" class="btn btn-default" (click)="addProductModal.show()">Add</button>
            </div>
          </div>
          <div class="col-md-6 part">
            <select formControlName="kind" class="form-control">
              <option [value]="null">Select component type</option>
              <option *ngFor="let kind of kindOptions" [value]="kind">
                {{ kind | titlecase }}
              </option>
            </select>
          </div>
          <div class="col-md-6">
            <textarea formControlName="hash" class="form-control" placeholder="Hash" (change)="validateHashIsHex($event.target.value)"></textarea>
            <!-- <button type="button" class="btn btn-default" (click)="hash2hex()">
              {{ hashToHexButtonText }}
            </button> -->
          </div>
          <div class="col-md-6">
            <textarea formControlName="description" class="form-control" placeholder="Description"></textarea>

          </div>
        </div>
      </div>
      <div class="col-md-5 right-part">
        <div class="flex-wrapper">
          <ng-select [items]="documents" bindLabel="label" bindValue="id" [multiple]="true"
            [closeOnSelect]="false" placeholder="Select linked components" formControlName="linkIds">
          </ng-select>
          <div class="error-box" *ngIf="displayErrors">

            <div class="error-message" *ngFor="let error of errors">
              <span class="z z-bolt"></span>{{ error }}
            </div>
          </div>
          <input type="submit" value="Register" class="btn btn-default">
        </div>
      </div>
    </div>
  </form>


  <dcs-modal #addProductModal title="Create a new project">
    <dcs-add-project [projects]="projects" (submitForm)="createProject($event); addProductModal.hide()"
      (cancel)="addProductModal.hide()" #addProject></dcs-add-project>
  </dcs-modal>

  <dcs-modal #registerModal title="Confirm new component">
    <dcs-confirm-document [document]="newDocument" [visible]="confirmVisible" (paymentSuccessful)="createComponent($event)"></dcs-confirm-document>

    <div class="footer">
      <button type="button" class="btn btn-default" (click)="registerModal.hide()">Cancel</button>
    </div>
  </dcs-modal>

</div>
