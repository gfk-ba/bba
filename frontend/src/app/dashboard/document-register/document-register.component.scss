@import '../../../styles/common/variables';
@import '../../../styles/common/zondicons';

textarea {
  margin-top: $default-margin;
  min-height: 5 * $default-margin;
  margin-bottom: $default-margin;
}

.left-part {
  .flex-wrapper {
    select {
      margin-right: 0.5 * $default-margin;
      flex-grow: 1;
    }
    button {
      min-width: 4rem;
    }
  }
}

.right-part {
  .flex-wrapper {
    height: 100%;
    flex-flow: column nowrap;
    justify-content: flex-start;
    input.btn {
      margin-top: auto;
      margin-left: auto;
    }
  }
  .error-box {
    .error-message {
      display: flex;
      align-items: center;
      margin: $default-margin 0 0 0;
      color: $red;
      .z {
        margin-right: 0.25em;
      }
    }
  }
}
