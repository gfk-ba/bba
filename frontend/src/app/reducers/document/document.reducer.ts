import { ActionReducer, combineReducers } from '@ngrx/store';
import { ICurrentDocumentState } from './current-document/current-document.manager';
import { currentDocument } from './current-document/current-document.reducer';
import { DocumentsListActions } from './documents-list/documents-list.actions';
import { IDocumentsListState } from './documents-list/documents-list.manager';
import { documentsList } from './documents-list/documents-list.reducer';

export interface State {
  currentDocument: ICurrentDocumentState;
  documentsList: IDocumentsListState;
}

export const reducer: ActionReducer<State, DocumentsListActions> = combineReducers({
  currentDocument,
  documentsList,
});
