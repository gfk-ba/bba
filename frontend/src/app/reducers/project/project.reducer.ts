import { ActionReducer, combineReducers } from '@ngrx/store';

import { ICurrentProjectState } from './current-project/current-project.manager';
import { currentProject } from './current-project/current-project.reducer';
import { IProjectsListState } from './projects-list/projects-list.manager';
import { projectsList } from './projects-list/projects-list.reducer';

export interface State {
  currentProject: ICurrentProjectState;
  projectsList: IProjectsListState;
}

export const reducer: ActionReducer<State> = combineReducers({
  currentProject,
  projectsList,
});
