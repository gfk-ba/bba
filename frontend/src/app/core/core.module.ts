import { registerLocaleData } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import localeDe from '@angular/common/locales/de';
import { InjectionToken, LOCALE_ID, NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { ServiceWorkerModule } from '@angular/service-worker';
import { APP_ENVIRONMENT } from '@dcs/ngx-tools';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { ActionReducerMap, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { ApiInterceptor } from './api.interceptor';
import { environment } from '../../environments/environment';
import { metaReducers, reducers, State } from '../reducers';
import { LoginGuard } from '../reducers/login/login.guard';

export const REDUCER_TOKEN = new InjectionToken<ActionReducerMap<State>>('Registered Reducers');

export function getReducers() {
  return reducers;
}

@NgModule({
  imports: [
    BrowserModule,
    HttpClientModule,
    StoreModule.forRoot(REDUCER_TOKEN, { metaReducers }),
    !environment.production ? StoreDevtoolsModule.instrument() : [],
    StoreRouterConnectingModule.forRoot({
      stateKey: 'router',
    }),
    ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production }),
  ],
  declarations: [],
  providers: [
    { provide: LOCALE_ID, useValue: 'en' },
    { provide: APP_ENVIRONMENT, useValue: environment },
    {
      provide: REDUCER_TOKEN,
      useFactory: getReducers,
    },
    {
      provide: HTTP_INTERCEPTORS,
      useClass: ApiInterceptor,
      multi: true,
    },
    LoginGuard,
  ],
})
export class CoreModule {
  constructor() {
    registerLocaleData(localeDe);
  }
}
