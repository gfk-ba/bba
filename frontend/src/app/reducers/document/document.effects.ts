import { Inject, Injectable } from '@angular/core';
import { APP_ENVIRONMENT } from '@dcs/ngx-tools';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { timer } from 'rxjs';
import { flatMap, mapTo, takeUntil } from 'rxjs/operators';
import { Environment } from '../../../environments/environment';
import {
  DocumentsListActionTypes,
  FetchDocumentsList,
} from './documents-list/documents-list.actions';

@Injectable()
export class DocumentEffects {
  @Effect()
  pollDocuments = this.actions$.pipe(
    ofType(DocumentsListActionTypes.StartPolling),
    flatMap(() =>
      timer(0, this.env.documentPollingTimer).pipe(
        takeUntil(
          this.actions$.pipe(
            ofType(DocumentsListActionTypes.StopPolling, DocumentsListActionTypes.StartPolling)
          )
        ),
        mapTo(new FetchDocumentsList())
      )
    )
  );

  constructor(private actions$: Actions, @Inject(APP_ENVIRONMENT) private env: Environment) {}
}
