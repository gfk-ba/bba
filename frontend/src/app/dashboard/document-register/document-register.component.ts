import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { ModalComponent } from '../../components/modal/modal.component';
import { Document, IDocument } from '../../reducers/document/models/document.class';
import { Project } from '../../reducers/project/models/project.class';
import { fromHex, isHex, toHex } from '../../utils/hex';
import {
  ChangeDetectionStrategy,
  Component,
  EventEmitter,
  Input,
  Output,
  ViewChild,
  ChangeDetectorRef,
} from '@angular/core';

@Component({
  selector: 'dcs-document-register',
  templateUrl: './document-register.component.html',
  styleUrls: ['./document-register.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class DocumentRegisterComponent {
  @Input()
  public documents: Document[];
  @Input()
  public projects: Project[];
  @Input()
  public kindOptions: string[];

  @Output()
  registerDocument = new EventEmitter<IDocument>();

  @Output()
  doCreateProject = new EventEmitter<any>();

  @ViewChild('registerModal')
  public readonly registerModal: ModalComponent;

  public form: FormGroup;
  public hashIsHex = true;
  public confirmVisible = false;
  public errors: string[] = [];

  public newDocument: Document | null;

  private submitted = false;
  private errorMessages: { [key: string]: string } = {
    projectId: 'Please select a project.',
    kind: 'Please select the kind.',
    hash: 'Please enter the hash value.',
    description: 'Please enter a description.',
  };

  constructor(fb: FormBuilder, private cd: ChangeDetectorRef) {
    this.form = fb.group({
      projectId: [null, Validators.required],
      kind: [null, Validators.required],
      hash: ['', [Validators.required, Validators.maxLength(256)]],
      description: ['', Validators.required],
      linkIds: [[]],
    });
  }

  get displayErrors(): boolean {
    return this.submitted && this.form.invalid;
  }

  get hashToHexButtonText(): string {
    return this.hashIsHex ? 'Hex-decode' : 'Hex-encode';
  }

  public submitForm(event?: Event) {
    if (event) {
      event.preventDefault();
    }

    this.submitted = true;
    this.errors = [];

    if (this.form.valid) {
      const values = this.form.value;
      const project = this.projects.find(d => String(values.projectId) === String(d.id)).toObject();
      const links = this.documents
        .filter(d => values.linkIds.includes(d.id))
        .map(d => d.toObject());

      this.newDocument = new Document({
        ...values,
        links,
        project,
      });

      this.registerModal.show();
      this.confirmVisible = true;
      this.cd.markForCheck();
    } else {
      Object.keys(this.form.controls).forEach((key: string) => {
        if (this.form.controls[key].errors) {
          this.errors.push(this.errorMessages[key]);
        }
      });
    }
  }

  public createComponent(payment: any) {
    if (this.form.valid) {
      this.registerDocument.next({ ...this.form.value, payment });
      this.form.reset({ linkIds: [] });
      this.registerModal.hide();
    }
  }

  public validateHashIsHex(value: string) {
    this.hashIsHex = isHex(value);
  }

  public hash2hex(): void {
    this.hashIsHex ? this.convertHashToText() : this.convertHashToHex();
  }

  public convertHashToHex() {
    const hashControl = this.form.get('hash');
    let value: string = hashControl.value;

    if (!isHex(value)) {
      value = toHex(value);
      this.validateHashIsHex(value);
      hashControl.setValue(value);
    }
  }

  public convertHashToText() {
    const hashControl = this.form.get('hash');
    let value: string = hashControl.value;

    if (isHex(value)) {
      value = fromHex(value);
      this.validateHashIsHex(value);
      hashControl.setValue(value);
    }
  }

  public createProject(data: { name: string; description: string }) {
    this.doCreateProject.next(data);
  }
}
