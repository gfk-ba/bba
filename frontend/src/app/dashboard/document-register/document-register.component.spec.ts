import { Component, NO_ERRORS_SCHEMA, ViewChild } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { NgSelectModule } from '@ng-select/ng-select';
import { DocumentRegisterComponent } from './document-register.component';
import { configureComponentTestSuite } from '../../../testing';
import { Document } from '../../reducers/document/models/document.class';
import { Project } from '../../reducers/project/models/project.class';

@Component({
  selector: 'dcs-test-component',
  template: `
    <dcs-document-register
      [documents]="documents"
      [projects]="projects"
    ></dcs-document-register>`,
})
class TestHostComponent {
  @ViewChild(DocumentRegisterComponent)
  public subject: DocumentRegisterComponent;

  public projects: Project[] = [new Project({ id: '42', name: 'Project 1' })];
  public documents: Document[] = [
    new Document({
      id: '1',
      projectId: 1,
      description: 'Component 1 of Project 1',
      kind: 'data',
      hash: '12345',
      address: '0xa3fb33a771b2e8863d27b155384b46b114455c92e7e199ba57b1be71568f06e8',
      linkIds: [],
      state: 'ok',
      stateDetails: '',
      createdAt: '2018-08-20T14:28:43.863Z',
    }),
    new Document({
      id: '2',
      projectId: 1,
      description: 'Component 2 of Project 1',
      kind: 'model',
      hash: '12345',
      address: '0xa3fb33a771b2e8863d27b155384b46b114455c92e7e199ba57b1be71568f06e8',
      linkIds: [],
      state: 'ok',
      stateDetails: '',
      createdAt: '2018-08-20T14:28:43.863Z',
    }),
    new Document({
      id: '3',
      projectId: 1,
      description: 'Component 3 of Project 1',
      kind: 'result',
      hash: '12345',
      address: '0xa3fb33a771b2e8863d27b155384b46b114455c92e7e199ba57b1be71568f06e8',
      linkIds: [1, 2],
      state: 'ok',
      stateDetails: '',
      createdAt: '2018-08-20T14:28:43.863Z',
    }),
  ];
}

describe('DocumentRegisterComponent', () => {
  let testHost: TestHostComponent;
  let fixture: ComponentFixture<TestHostComponent>;
  let subject: DocumentRegisterComponent;
  let subjectEl: HTMLUnknownElement;

  configureComponentTestSuite();

  beforeAll(async () => {
    await TestBed.configureTestingModule({
      declarations: [TestHostComponent, DocumentRegisterComponent],
      imports: [ReactiveFormsModule, NgSelectModule],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    testHost = fixture.componentInstance;
    subject = testHost.subject;
    subjectEl = fixture.nativeElement.querySelector('dcs-document-register');
    fixture.detectChanges();
  });

  it('should create testHost', () => {
    expect(testHost).toBeTruthy();
  });

  it('should render component', () => {
    expect(subjectEl).toBeInstanceOf(HTMLUnknownElement);
    expect(subject).toBeInstanceOf(DocumentRegisterComponent);
  });

  describe('submitForm', () => {
    let registerModalShowSpy: jest.Mock;

    const formData = {
      projectId: '42',
      kind: 'custom',
      hash: '12345',
      description: 'Desc',
      linkIds: [1, 2, 3],
    };

    beforeEach(() => {
      subject.registerModal.show = () => {};
      registerModalShowSpy = jest.spyOn(subject.registerModal, 'show').mockImplementation(() => {});
      subject.form.setValue(formData);

      subject.submitForm(new CustomEvent('submit'));
    });

    it('shows the confirm modal', () => {
      expect(registerModalShowSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('createComponent', () => {
    let resetSpy: jest.SpyInstance;
    let registerDocumentSpy: jest.SpyInstance;
    let registerModalHideSpy: jest.Mock;

    const formData = {
      projectId: '42',
      kind: 'custom',
      hash: '12345',
      description: 'Desc',
      linkIds: [1, 2, 3],
    };

    beforeEach(() => {
      resetSpy = jest.spyOn(subject.form, 'reset');
      registerDocumentSpy = jest.spyOn(subject.registerDocument, 'next');
      subject.registerModal.hide = () => {};
      registerModalHideSpy = jest.spyOn(subject.registerModal, 'hide').mockImplementation(() => {});

      subject.form.setValue(formData);
      subject.createComponent({ coupon: 'test' });
    });

    it('resets the form', () => {
      expect(resetSpy).toHaveBeenCalledTimes(1);
      expect(resetSpy).toHaveBeenCalledWith({ linkIds: [] });
      expect(subject.form.value).toMatchSnapshot();
    });

    it('fires the EventEmitter', () => {
      expect(registerDocumentSpy).toHaveBeenCalledWith({
        ...formData,
        payment: { coupon: 'test' },
      });
    });

    it('closes the modal', () => {
      expect(registerModalHideSpy).toHaveBeenCalledTimes(1);
    });
  });
});
