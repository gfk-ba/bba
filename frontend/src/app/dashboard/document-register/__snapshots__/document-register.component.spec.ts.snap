// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentRegisterComponent createComponent resets the form 1`] = `
Object {
  "description": null,
  "hash": null,
  "kind": null,
  "linkIds": Array [],
  "projectId": null,
}
`;
